api: Fix return from `glfs_h_open()` to honour `O_DIRECTORY` flag

anoopcs9 opened the pull request:
This is more or less the same change done previously for _glfs_open()_. See description from https://github.com/gluster/glusterfs/pull/3307 and corresponding commit https://github.com/gluster/glusterfs/commit/754a8a8e8ec8ea3a9949ecfd4e3ce149e0bd99b8 for more details.

anoopcs9 commented:
/run regression

amarts reviewed (commented):
LGTM

anoopcs9 commented:
@xhernandez can we take this in?

xhernandez commented:
The change seems correct, but I would prefer to move all the common logic into a single function and call it from `glfs_open()` and `glfs_h_open()`. This way we'll minimize the chances of future inconsistencies between these functions if another modification is made.

anoopcs9 committed:
gfapi: Add a helper function to validate open flags

This is in preparation to be invoked commonly via glfs_open() and
glfs_h_open().

Signed-off-by: Anoop C S <anoopcs@cryptolab.net>

anoopcs9 committed:
gfapi: Use helper function in glfs_open() for flags validation

Signed-off-by: Anoop C S <anoopcs@cryptolab.net>

anoopcs9 committed:
gfapi: Fix return from glfs_h_open() to honour O_DIRECTORY flag

This is more or less the same change done previously for glfs_open().
See commit message from 754a8a8e8ec8ea3a9949ecfd4e3ce149e0bd99b8 for
more details.

Signed-off-by: Anoop C S <anoopcs@cryptolab.net>

anoopcs9 commented:
> The change seems correct, but I would prefer to move all the common logic into a single function and call it from `glfs_open()` and `glfs_h_open()`. This way we'll minimize the chances of future inconsistencies between these functions if another modification is made.

I have a small function to handle open flags now in the updated version.

anoopcs9 commented:
/run regression

xhernandez commented on api/src/glfs.c:
```diff
@@ -642,6 +642,31 @@ unset_fop_attr(dict_t **fop_attr)
+        if (flags & O_DIRECTORY) {
+            ret = -1;
+            errno = ENOTDIR;
+        }
+        if (!IA_ISREG(ia_type)) {
```
```suggestion
        } else if (!IA_ISREG(ia_type)) {
```

anoopcs9 replied:
Good catch.. copy/paste sometimes comes with such caveats 😛

anoopcs9 commented:
/run regression
